mlflow/pytorch/pytorch_autolog.py

```diff
@@ -0,0 +1,197 @@
+# Eventhough the same can be achieved using
+# MlflowLogger(Pytorch Lightning's inbuild class),
+# following are the downsides in using MlflowLogger.
+# 1. MlflowLogger doesn't provide a mechanism to dump an entire model into mlflow.
+#    Only model checkpoint is saved.
+# 2. For dumping the model into mlflow `mlflow.pytorch` library is used
+# and the library expects `mlflow` object to be instantiated.
+# In case of MlflowLogger, Run management is completely controlled by the class and
+# hence mlflow object needs to be reinstantiated by setting
+# tracking uri, experiment_id and run_id which may lead to a race condition.
+
+
+def autolog(log_every_n_iter=1):
+    global every_n_iter
+    every_n_iter = log_every_n_iter
+
+    class __MLflowPLCallback(pl.Callback):
+        """
+        Callback for auto-logging metrics and parameters.
+        """
+
+        def __init__(self):
+            self.early_stopping = False
+
+        def on_epoch_end(self, trainer, pl_module):
+            """
+            Log loss and other metrics values after each epoch
+
+            :param trainer: pytorch lightning model instance
+            :param pl_module: pytorch lightning base module
+            """
+            if (pl_module.current_epoch + 1) % every_n_iter == 0:
+                for key, value in trainer.callback_metrics.items():
+                    try_mlflow_log(
+                        mlflow.log_metric, key, float(value + 1), step=pl_module.current_epoch
+                    )
+
+            for callback in trainer.callbacks:
+                if isinstance(callback, pl.callbacks.early_stopping.EarlyStopping):
+                    self._early_stop_check(callback)
+
+        def on_train_start(self, trainer, pl_module):
+            """
+            Logs Optimizer related metrics when the train begins
+
+            :param trainer: pytorch lightning model instance
+            :param pl_module: pytorch lightning base module
+            """
+            mlflow.set_tag(key="Mode", value="training")
+            try_mlflow_log(mlflow.log_param, "epochs", trainer.max_epochs)
+
+            for callback in trainer.callbacks:
+                if isinstance(callback, pl.callbacks.early_stopping.EarlyStopping):
+                    self.early_stopping = True
+                    self._log_early_stop_params(callback)
+
+            if hasattr(trainer, "optimizers"):
+                for optimizer in trainer.optimizers:
+                    try_mlflow_log(mlflow.log_param, "optimizer_name", type(optimizer).__name__)
+
+                    if hasattr(optimizer, "defaults"):
+                        defaults_dict = optimizer.defaults
+                        if "lr" in defaults_dict:
+                            try_mlflow_log(mlflow.log_param, "learning_rate", defaults_dict["lr"])
+
+                        if "eps" in defaults_dict:
+                            try_mlflow_log(mlflow.log_param, "epsilon", defaults_dict["eps"])
```

Comment:
Why not log all parameters stored in `defaults_dict`?

Reply:
@harupy The default names were cryptic. So changed to full name for better readability and ease of use during metrics comparison in the MLflow UI